### PR TITLE
Make daemon require a workspace load target, so that daemons can fetch the set of available locations in the workspace

### PR DESCRIPTION
**Pull request by gibsondan**
This is a slightly breaking change that I'd like to pitch as part of the 0.14.0 release - it's arguably how the daemon should have worked from the beginning, and will unblock features like automatically turning on schedules and sensors based on code.

Right now, the daemon has no concept of the workspace in the same way that dagit does. The only way to, say, check for running schedules is to rely on there being rows in the schedule DB.

This PR requires the same arguments to the `dagster-daemon run` command that are currently passed into Dagit to specify the workspace, and creates a new workspace class that reads from that workspace just like dagit does. There's no need to restart the daemon when the workspace changes (just like dagit) - new locations can be picked up automatically by clearing the workspace and reloading things.

This will not be a breaking change for people using the Helm chart, but folks who are running the daemon locally or in Docker may need to take steps to ensure that the same workspace.yaml is being used in the daemon pod or container. If we want to do this, a separate PR will update the docs and examples to make sure that they all configure the daemon correctly[

As of this PR, there are no actual functionality changes - just a new method that daemons can use to get access to the full workspace.

### CODE REVIEW

**Comment by gibsondan** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6074** 
    * **PR #6077**  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

**Review by alangenfeld — COMMENTED** [review]
Having a hard time thinking of a reason not to do this.

Well definitely want a really clear error message in the `dagster-daemon` cli entry when the information is missing.

I think its good that this makes it more clear what the daemon process needs to have access to, as opposed to it being indirected through values in the database.

**Comment by prha** [comment]
I also think this is a really good change, and something we should land for `0.14.0`...

File: python_modules/dagster/dagster/core/host_representation/origin.py [hunk 1]
```diff
@@ -197,7 +197,13 @@ def create_location(self) -> NoReturn:
 
     @contextmanager
     def create_test_location(self):
-        from dagster.core.workspace.dynamic_workspace import DynamicWorkspace
+        """Back-compat after create_test_location was renamed to create_single_location"""
+        with self.create_single_location() as location:
+            yield location
+
+    @contextmanager
+    def create_single_location(self):
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/host_representation/origin.py:205]: bit confused about this - maybe worth some more exposition in the comment or double check it makes sense

File: python_modules/dagster/dagster/core/workspace/dynamic_workspace.py [hunk 2]
```diff
@@ -1,70 +0,0 @@
-from dagster import check
-from dagster.core.host_representation.origin import RepositoryLocationOrigin
-from dagster.core.host_representation.repository_location import GrpcServerRepositoryLocation
-from dagster.core.workspace import IWorkspace
-
-
-class DynamicWorkspace(IWorkspace):
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/workspace/dynamic_workspace.py:7]: bye bye

File: python_modules/dagster/dagster/daemon/workspace.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            load_error=error,
+            load_status=WorkspaceLocationLoadStatus.LOADED,
+            display_metadata=location.get_display_metadata()
+            if location
+            else origin.get_display_metadata(),
+            update_timestamp=time.time(),
+        )
+
+    def _create_location_from_origin(self, origin):
+        check.inst_param(origin, "origin", RepositoryLocationOrigin)
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/daemon/workspace.py:121]: nit: mypy types be nice

**Review by alangenfeld — APPROVED** [review]
code looks fine by me - do you want to include required documentation updates in this diff or subsequent?

**Pull request merged**